src/alvin0319/OffHand/OffHandInventory.php

```diff
@@ -27,75 +27,57 @@
-use pocketmine\Player;
-
-class OffHandInventory extends BaseInventory{
-	public const SLOT_OFFHAND = 0;
-
-	/** @var Player */
+use pocketmine\network\mcpe\protocol\types\ContainerIds;
+use pocketmine\nbt\tag\ListTag;
+use pocketmine\item\Item;
+use pocketmine\entity\Human;
+use pocketmine\Server;
+use pocketmine\utils\MainLogger;
+class OffhandInventory extends BaseInventory{
+    /** @var Human */
 	protected $holder;
-
-	public function __construct(Player $holder){
-		parent::__construct([], 1);
+	public function __construct(Human $holder){
 		$this->holder = $holder;
+		parent::__construct();
 	}
-
-	public function getPlayer() : Player{
-		return $this->holder;
-	}
-
-	public function setSize(int $size) : void{
-		throw new BadMethodCallException("Cannot call setSize on OffHandInventory");
-	}
-
 	public function getName() : string{
-		return "OffHandInventory";
+		return "Offhand";
 	}
-
 	public function getDefaultSize() : int{
 		return 1;
 	}
-
-	public function setItemInOffHand(Item $item) : void{
-		$this->setItem(self::SLOT_OFFHAND, $item);
-		
-		$pk = new InventoryContentPacket();
-		$pk->windowId = $this->holder->getWindowId($this);
-		$pk->items = [ItemStackWrapper::legacy($item)];
-		$this->holder->sendDataPacket($pk);
-
-		$this->holder->getDataPropertyManager()->setByte(Entity::DATA_COLOR, Entity::DATA_TYPE_BYTE);
-
-		$this->sendMobEquipmentPacket($this->holder);
-
-		foreach($this->viewers as $player){
-			$this->sendMobEquipmentPacket($player);
-		}
+	public function setSize(int $size){
+		throw new \BadMethodCallException("Offhand can only carry one item at a time");
 	}
-
-	public function getItemInOffHand() : Item{
-		return $this->getItem(self::SLOT_OFFHAND);
+	public function setItem(int $index, Item $item, bool $send = true): bool{
+	    	parent::setItem($index, $item, $send);
+	    	$this->sendItem();
+	    	$this->sendContents($this->getHolder());
+		$this->getHolder()->namedtag->setTag(new ListTag("Offhand", [$item->nbtSerialize()]));
+		return true;
 	}
-
-	public function sendMobEquipmentPacket(Player $player) : void{
-		$pk = new MobEquipmentPacket();
-		$pk->windowId = $this->getPlayer()->getWindowId($this);
-		$pk->item = $this->getItemInOffHand();
-		$pk->entityRuntimeId = $this->getPlayer()->getId();
-		$pk->hotbarSlot = $pk->inventorySlot = 0;
-		$player->sendDataPacket($pk);
-	}
-
-	public function onSlotChange(int $index, Item $before, bool $send) : void{
-		foreach($this->viewers as $player){
-			$this->sendMobEquipmentPacket($player);
+	public function sendItem(array $players = null): void{
+	    	$players = $players ?? $this->holder->getViewers();
+		$pk = new MobEquipmentPacket;
+		$pk->windowId = ContainerIds::OFFHAND;
+		$pk->item = $this->getItem(0);
+		$pk->inventorySlot = $pk->hotbarSlot = 0;
+		$pk->entityRuntimeId = $this->getHolder()->getId();
+		foreach($players as $player){
+			try{
+		    		$player->batchDataPacket($pk);
```

Comment:
what is the purpose of this? you can just use `$player->sendDataPacket()`

Reply:
When I used  `$player->sendDataPacket()` it wouldn’t work, so I had to use `$player->batchDataPacket()` instead and worked perfectly

Reply:
Also I added `try{}` in case if someone uses your plugin and do `$offhand->sendItem([$player])` and the `$player` var isn’t really a player or a object at all, so it wouldn’t crash, will just notify that error